File: scala/Cell.scala
import scala.collection.mutable.ArrayBuffer

class Cell(
  val x: Int,
  val y: Int,
  var alive: Boolean = false,
):
  var nextState: Option[Boolean] = None
  val neighbours = ArrayBuffer[Cell]()

  def toChar =
    if alive then "o" else " "

  def aliveNeighbours =
    // The following is slower
    // neighbours.filter(_.alive).length

    // The following is slower
    // var aliveNeighbours = 0
    // for neighbour <- neighbours do
    //   if neighbour.alive then
    //     aliveNeighbours += 1
    // aliveNeighbours

    // The following is the fastest
    var aliveNeighbours = 0
    val count = neighbours.length
    for i <- 0 until count do
      val neighbour = neighbours(i)
      if neighbour.alive then
        aliveNeighbours += 1
    aliveNeighbours

File: ruby/cell.rb
class Cell
  attr_reader :x, :y
  attr_accessor :alive, :next_state, :neighbours

  def initialize(x, y, alive = false)
    @x = x
    @y = y
    @alive = alive
    @next_state = nil
    @neighbours = []
  end

  def to_char
    @alive ? 'o' : ' '
  end

  def alive_neighbours()
    # The following is the fastest
    neighbours.count(&:alive)

    # The following is slower
    # alive_neighbours = 0
    # neighbours.each do |neighbour|
    #   alive_neighbours += 1 if neighbour.alive
    # end
    # alive_neighbours

    # The following is slower
    # alive_neighbours = 0
    # count = neighbours.size
    # for i in 0...count do
    #   neighbour = neighbours[i]
    #   alive_neighbours += 1 if neighbour.alive
    # end
    # alive_neighbours
  end
end
